ext/win32ole/win32ole.c

```diff
@@ -812,23 +815,52 @@ ole_initialize(void)
     HRESULT hr;
 
     if(!g_uninitialize_hooked) {
-	rb_add_event_hook(ole_uninitialize_hook, RUBY_EVENT_THREAD_END, Qnil);
-	g_uninitialize_hooked = TRUE;
+    rb_add_event_hook(ole_uninitialize_hook, RUBY_EVENT_THREAD_END, Qnil);
+    g_uninitialize_hooked = TRUE;
     }
 
     if(g_ole_initialized == FALSE) {
-        hr = OleInitialize(NULL);
+        if(running_nano_server() == TRUE) {
+            hr = CoInitializeEx(NULL, COINIT_MULTITHREADED);
+        } else {
+            hr = OleInitialize(NULL);
+        }
         if(FAILED(hr)) {
             ole_raise(hr, rb_eRuntimeError, "fail: OLE initialize");
         }
         g_ole_initialized_set(TRUE);
 
-        hr = CoRegisterMessageFilter(&imessage_filter, &previous_filter);
-        if(FAILED(hr)) {
-            previous_filter = NULL;
-            ole_raise(hr, rb_eRuntimeError, "fail: install OLE MessageFilter");
+        if(running_nano_server() == FALSE) {
+            hr = CoRegisterMessageFilter(&imessage_filter, &previous_filter);
+            if(FAILED(hr)) {
+                previous_filter = NULL;
+                ole_raise(hr, rb_eRuntimeError, "fail: install OLE MessageFilter");
+            }
+        }
+    }
+}
+
+static BOOL
+running_nano_server(void)
+{
+    HKEY hsubkey;
+    LONG err;
+    VALUE subkey;
+
+    if(g_nano_checked == FALSE) {
+        subkey = rb_str_new2("SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Server\\ServerLevels");
```

Comment:
I think in your example `rb_str_new2` and `StringValuePtr` is used because of following lines

```c
rb_str_concat(subkey, com);
rb_str_cat2(subkey, "\\CLSID");
```

where `rb_str_*` functions are used to manipulate `subkey`.

I think you can use just `const char *` to avoid useless typing.